project/src/pv_speaker.c

```diff
@@ -0,0 +1,437 @@
+/*
+    Copyright 2024 Picovoice Inc.
+
+    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
+    file accompanying this source.
+
+    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations under the License.
+*/
+
+#pragma GCC diagnostic push
+
+#pragma GCC diagnostic ignored "-Wunused-result"
+
+#define MINIAUDIO_IMPLEMENTATION
+
+#include "miniaudio.h"
+
+#pragma GCC diagnostic pop
+
+#include "pv_circular_buffer.h"
+#include "pv_speaker.h"
+
+#define PV_SPEAKER_DEFAULT_DEVICE_INDEX (-1)
+
+#define PV_SPEAKER_VERSION "1.0.0"
+
+static const int32_t WRITE_RETRY_COUNT = 500;
+static const int32_t WRITE_SLEEP_MILLI_SECONDS = 2;
+
+static bool is_empty = false;
+static bool is_data_requested_while_empty = false;
+
+struct pv_speaker {
+    ma_context context;
+    ma_device device;
+    pv_circular_buffer_t *buffer;
+    int32_t frame_length;
+    bool is_started;
+    bool is_debug_logging_enabled;
+    ma_mutex mutex;
+};
+
+static void pv_speaker_ma_callback(ma_device *device, void *output, const void *input, ma_uint32 frame_count) {
+    (void) input;
+
+    pv_speaker_t *object = (pv_speaker_t *) device->pUserData;
+
+    ma_mutex_lock(&object->mutex);
+    int32_t read_length = 0;
+    pv_circular_buffer_read(object->buffer, output, (int32_t) frame_count, &read_length);
+    if (object->is_debug_logging_enabled) {
+        fprintf(stdout, "[INFO] Number of frames copied to output buffer: %d.\n", read_length);
```

Comment:
lets just print a warning only if the `read_length` is different from the requested one.

Reply:
I removed the log - realized that `read_length` can still return 0 right after `start()`, since it starts reading right away before anything is written. This was confusing since `is_empty` is more like `is_stopped_and_empty`.